scene/resources/visual_shader_nodes.cpp

```diff
@@ -4936,8 +4936,11 @@ void VisualShaderNodeVectorCompose::set_op_type(OpType p_op_type) {
 		default:
 			break;
 	}
+	bool port_count_changed = p_op_type < op_type;
 	op_type = p_op_type;
-	emit_changed();
+	if (port_count_changed) {
+		emit_changed();
+	}
```

Comment:
This is incorrect for several reasons. First of all, the signal means that the resource has been changed for any reason, which is always true by this line since we rule out the new and the old op type being the same at the start of this method.

And second, your proposed change assumes that the order of enum values would somehow correspond to the number of available ports. While it's true in this case, it's a weak relation and such code can very easily lead to annoying mistakes and bugs. We could discuss the merits of it, but as mentioned above there is no need for this change at all.

Reply:
The check here is not incorrect, although admittedly I should have chosen another name for the boolean. What I am testing specifically is whether the number of ports has decreased, not simply changed, because we do not need to do anything when the port count increased and we do not have any way to know which type of change it was when handing the signal

Reply:
Well, the signal is there to indicate that the resource has changed, whether the number of ports increased, decreased, or remained the same, it's not relevant. The signal is emitted because a new value for the op type was set.

Reply:
To add a new signal you need to use the `ADD_SIGNAL` macro in the `_bind_methods` method of the class that you want to add it to. In this case that would probably be `VisualShaderNode`. Then just emit it with `emit_signal("your_signal_name")` and connect to it normally.